synapse/tools/migrate_020.py

```diff
@@ -1225,23 +1248,34 @@ class Migrator(s_base.Base):
         nodeedits = []
         t_strt = s_common.now()
         stot = 0
+        fastfwd = fromlast  # so we can stop the comparison once caught up
         async for node in self._srcIterNodes(src_slab, src_bybuid):
+            stot += 1
+
+            if stot % 1000000 == 0:  # pragma: no cover
+                logger.info(f'...on node {stot:,} for layer {iden}')
+
+            if stot % fairiter == 0:
+                await asyncio.sleep(0)
+
+            if fastfwd:
+                if stot < startfrom:
+                    continue
+                else:
+                    fastfwd = False
+
             buid = node[0]
             form = node[1]['ndef'][0].replace('*', '')
             src_fcnt[form] += 1
 
-            stot += 1
             if nodelim is not None and stot >= nodelim:
                 logger.warning(f'Stopping node migration due to reaching nodelim {stot}')
                 # checkpoint is the next node to add
                 await self._migrlogAdd(migrop, 'chkpnt', iden, (buid, stot, s_common.now()))
                 break
 
-            if stot % 1000000 == 0:  # pragma: no cover
-                logger.info(f'...on node {stot:,} for layer {iden}')
-
-            if stot % fairiter == 0:
-                await asyncio.sleep(0)
+            if stot % savechkpnt == 0:
+                await self._migrlogAdd(migrop, 'chkpnt', iden, (buid, stot, s_common.now()))
```

Comment:
Maybe you want to save a checkpoint when you get to the end too?

Reply:
updated